### ISSUE TITLE
Discussion: upgrading to C++17

### ISSUE THREAD

**Issue by dongcarl** [bitcoin/bitcoin#16684]
**UPDATE**: Release plan summary from April 2nd, 2020 meeting:

|       | C++11 Compat       | C++17 Compat       | Gitian/Guix Release |
|-------|--------------------|--------------------|---------------------|
| v0.20 | :heavy_check_mark: | if easy            | C++11               |
| v0.21 | :heavy_check_mark: | :heavy_check_mark: | C++17               |
| v0.22 | :x:                | :heavy_check_mark: | C++17               |

Meeting logs can be found here: http://www.erisian.com.au/meetbot/bitcoin-core-dev/2020/bitcoin-core-dev.2020-04-02-19.01.log.html

-----

There were some discussion around upgrading to C++17. It seems to be able to give us library and language features that will simplify future code. ~~For example, for library features, our current `optional.h` is just a wrapper around `boost::optional`, and can be replaced with `std::optional`. Other library features that might be useful include (non-exhaustively): `std::variant`, the file system library, and `std::shared_mutex`.~~ @elichai has kindly provided a table summarizing features useful for Bitcoin Core here: https://github.com/bitcoin/bitcoin/issues/16684#issuecomment-524067463.

I want to apologize for misspeaking about C++17 _requiring_ GCC8, that is **untrue**, and I believe all of the features we would want will be covered by GCC7, and some by GCC6.

For some background on the **cons**: https://github.com/bitcoin/bitcoin/issues/13356#issuecomment-457870907

-----

Default GCC versions major distros (as of Aug 22nd):

|   | Operating System | release version | GCC version | Notes |
|---|------------------|--------------------------|-------------|------|
| :heavy_check_mark: | Debian           | Buster                   | 8.3         |      |
| :heavy_check_mark: | Ubuntu           | 18.04 LTS                | 7.4         |      |
| :heavy_check_mark: | Fedora           | 30                       | 9.1.1       |      |
| :heavy_check_mark: | RHEL             | 8                        | 8.2         |      |
|  | CentOS           | 7                        | 4.8.5       | CentOS 8 (GCC 8.2) is being released [as we speak](https://wiki.centos.org/About/Building_8#head-516d5e6556bb8523b52fba246953d32831582480) |

-----

In-depth code examples of

C++17 features: https://github.com/AnthonyCalandra/modern-cpp-features/blob/master/CPP17.md
C++14 features: https://github.com/AnthonyCalandra/modern-cpp-features/blob/master/CPP14.md

-----

Here are some compatibility tables from https://en.cppreference.com/w/cpp/compiler_support#cpp17

Library features:
![Screen Shot 2019-08-22 at 19 39 24](https://user-images.githubusercontent.com/3445290/63544344-91f0b580-c514-11e9-910b-64b82d93ad9d.png)

Language features:
![Screen Shot 2019-08-22 at 19 38 32](https://user-images.githubusercontent.com/3445290/63544316-7b4a5e80-c514-11e9-8ce1-4cf873127536.png)

**Comment by practicalswift** [comment]
Some C++14 features worth mentioning as well: `std::make_unique`, more expressive `constexpr` functions allowed, return type deduction (`auto foo() { … }`), `[[deprecated]]`, binary literals (`0b10101110`) and polymorphic lambdas (`[](auto foo) { … }`).

**Comment by elichai** [comment]
| Name                              | Important | Nice to have | Boost Replacement |   |
|-----------------------------------|-----------|--------------|-------------------|---|
| std::variant                      |           |              | V                 |   |
| std::make_unique                  |           | V            |                   |   |
| Variable Templates                |           | V            |                   |   |
| Conditions in constexpr functions |           | V            |                   |   |
| auto in lambdas params            |           | V            |                   |   |
| Binary literals                   |           | VV           |                   |   |
| std::filesystem                   | V         |              | V                 |   |
| std::optional                     |           |              | V                 |   |
| [[nodiscard]]+[[fallthrough]]     |           | V            |                   |   |
| de-structuring initialization     |           | V            |                   |   |

**Comment by elichai** [comment]
Another reason we might want to upgrade:
very powerful const expressions. both c++14 and 17 made `constexpr` way more powerfull.
such that we could start making a lot of functions `constexpr` which might **maybe** result in some performance increase (for functions that might be able to be evaluated at compile time)

**Comment by MarcoFalke** [comment]
Is C++17 supported on Ubuntu Xenial, Debian 9, and the *BSD operating systems?

**Comment by ysangkok** [comment]
Debian 9 ships GCC 6.3 according to [the development packages listing](https://packages.debian.org/oldstable/devel/). GCC 6 is [listed as supporting some features of C++17 experimentally](https://gcc.gnu.org/gcc-6/changes.html).

Ubuntu Xenial ships with a [default GCC 5.3](https://packages.ubuntu.com/xenial/gcc), which is even older. A PPA offers newer GCC versions: [~ubuntu-toolchain-r](https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test?field.series_filter=xenial)

**Comment by laanwj** [comment]
Added 0.21 milestone optimistically

**Comment by sanjaykdragon** [comment]
I agree with moving to C++17, but when do we plan to move to the soon coming C++20?

**Comment by sipa** [comment]
@sanjaykdragon It's not that easy, as we need to remain compatible with various stable Linux distributions, which don't update their C++ compilers that frequently (or at all, after release of a particular distro version).

C++20 does not exist yet, so that seems entirely out of scope.

**Comment by sanjaykdragon** [comment]
> @sanjaykdragon It's not that easy, as we need to remain compatible with various stable Linux distributions, which don't update their C++ compilers that frequently (or at all, after release of a particular distro version).
> 
> C++20 does not exist yet, so that seems entirely out of scope.

Are we just waiting for linux distributions to become compatible with C++17 for a migration?

**Comment by practicalswift** [comment]
Of the 13 build jobs we have in Travis it looks like three build jobs use compiler without C++17 support:  CentOS 7, Ubuntu 16.04 (Xenial) and macOS Sierra (10.12). Sounds correct? :)

**Comment by laanwj** [comment]
> Are we just waiting for linux distributions to become compatible with C++17 for a migration?

Yes, that's the main factor holding it up.

**Comment by sanjaykdragon** [comment]
Just a comment / improvement to make when we move to C++17:
use std::string_view more.

**Comment by practicalswift** [comment]
@sanjaykdragon `std::string_view` has _really_ sharp edges -- some would even claim it encourages use-after-free bugs. I think our recommendation should be to **avoid** `std::string_view` in the general case :)

**Comment by sanjaykdragon** [comment]
> @sanjaykdragon `std::string_view` has _really_ sharp edges -- some would even claim it encourages use-after-free bugs. I think our recommendation should be to **avoid** `std::string_view` in the general case :)

I'll leave this discussion until we actually start moving to C++17, but I think string_view would bring a lot of benefits if we do use it

**Comment by dongcarl** [comment]
Updated issue description with release plan summary from April 2nd, 2020 meeting.

**Comment by JeremyRubin** [comment]
To clarify; there is a difference between modules needed "compatibility level" without requiring linking in boost. A good rule of thumb is if in your module boost is already linked, then it would be OK to use a backport shim in it. This rules out libconsensus afaik; and is relevant to e.g. the work on Span<T>.

I don't have a good understanding of who is using libconsensus to comment on how bad it would be to require a backport to either use ++17 or link in boost, so I'll demur on making a policy there.

**Comment by sipa** [comment]
FWIW, with #18468, it seems the codebase is entirely C++17 compilable.

**Comment by MarcoFalke** [comment]
We should also look into bumping glibc from 2.17 to 2.18 at the same time: https://github.com/bitcoin/bitcoin/pull/18709/files

**Comment by fanquake** [comment]
Annoyingly, if we want to do this for 0.21.0, we'll also likely have to bump our minimum required macOS from 10.12 to 10.14. 

You cannot use `std::get` with `std::variant` on macOS < 10.14, because Apples libc++ doesn't support the `std::bad_variant_access` exception. [Relevant comment](https://github.com/bitcoin/bitcoin/pull/19183#discussion_r439794318) in #19183.

```cpp
#include <iostream>
#include <variant>

int main() {
	std::variant<int, std::string> i, s;
	i = 42;
	s = "Hello";
	try {
		std::cout << std::get<int>(i) << "\n";
	} catch (const std::bad_variant_access&) {}
}
```

```bash
clang++ --version                                               
Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin19.5.0
```
```bash
clang++ -std=c++17 std_variant.cpp -O2 -mmacos-version-min=10.13
std_variant.cpp:9:21: error: 'get<int, int, std::__1::basic_string<char> >' is unavailable: introduced in macOS 10.14
                std::cout << std::get<int>(i) << "\n";
                                  ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/variant:1376:16: note: 'get<int, int, std::__1::basic_string<char> >' has been explicitly marked unavailable here
constexpr _Tp& get(variant<_Types...>& __v) {
               ^
std_variant.cpp:10:22: error: 'bad_variant_access' is unavailable: introduced in macOS 10.14
        } catch (const std::bad_variant_access&) {}
                            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/variant:221:69: note: 'bad_variant_access' has been explicitly marked unavailable here
class _LIBCPP_EXCEPTION_ABI _LIBCPP_AVAILABILITY_BAD_VARIANT_ACCESS bad_variant_access : public exception {
                                                                    ^
2 errors generated.
```

While we could work around this in our own code, using `std::get_if`, this would still be a problem for 3rd-party dependencies. 

I've been testing Qt 5.15LTS (we'll have to enable c++17 in qt, and may upgrade to a newer version at the same time), and you can't enable `-std c++17`, while targeting a macOS deployment version < 10.14, configuring will fail. They are making use of `std::get` with `std::variant` throughout their cocoa code.

There is a similar problem with using `std::visit`. See my comment [here](https://github.com/bitcoin/bitcoin/pull/19183#discussion_r439794419).

**Comment by MarcoFalke** [comment]
For 0.21 it will simply be gitian/guix with C++17 on a C++11 codebase and that shouldn't affect macos, right? The actual bump will happen in 0.22, which is released not before next year.

**Comment by fanquake** [comment]
Sorry if I'm being slow, but I've re-read the discussion here, and I think I must be misunderstanding something.

> For 0.21 it will simply be gitian/guix with C++17 on a C++11 codebase 

What does "with C++17 on a C++11 codebase" mean? Are we requiring C++17 compatible compilers, and using them for releases (passing `--enable-c++17`?), but not actually using any C++17 features in the code?

For 0.21.0, are we compiling our dependencies, like Qt, in C++17 or [C++11 mode](https://github.com/bitcoin/bitcoin/blob/master/depends/packages/qt.mk#L27)? 

I'm not sure if we are planning on bumping to a newer Qt LTS (5.12 or 5.15) for 0.21.0. However I don't think we can do that if we want to remain C++11 compatible. I mentioned [here](https://github.com/bitcoin/bitcoin/issues/17768#issue-539394098), that Qt 5.12 has started using C++14 features in some portions of it's codebase, and essentially "forgotten" that 5.12.x is meant to be compatible/compilable with C++11.  Maybe it matters less for the GUI, but I haven't seen that discussion anywhere.

> The actual bump will happen in 0.22

Ok. So PRs like #19183 are just being opened quite a few months in advance of when they can be merged into master? We couldn't merge that PR as is, if we want to produce binaries that will run on macOS < 10.14 for 0.21.0. Using `std::get` with `std::variant`, and `std::visit`, requires a newer Apple libc++ dylib, that only began shipping with macOS 10.14.

**Comment by MarcoFalke** [comment]
> For 0.21.0, are we compiling our dependencies, like Qt, in C++17

No, everything (including depends) will stay at C++11. (Well except libmultiprocess, but that is an experimental package not compiled by default in any way)

> So PRs like #19183 are just being opened quite a few months in advance

Yes. I opened it so that any issues with C++17 can be detected upfront. Also, this pull allows other boost related refactoring pulls to be closed (e.g. #17953)

**Comment by MarcoFalke** [comment]
Actually I am not sure if depends should stay at C++11. If it is required to be bumped to C++17 to enable C++17 in core, then so be it. Is there any downside? cc @dongcarl

**Comment by fanquake** [comment]
> Actually I am not sure if depends should stay at C++11. If it is required to be bumped to C++17 to enable C++17

We can’t bump Qt to a version that has a C++17 mode, without breaking support for C++11. 

Bumping to C++17 would also require dropping support for macOS < 10.14.

I’ve outlined all of this in my comments in #19305.

**Comment by MarcoFalke** [comment]
>  dropping support for macOS < 10.14

Isn't that caused by bumping qt? In theory (not saying that we should), depends and the configure flags could be bumped to enable C++17 without touching anything else in gitian/guix and (hopefully) without breaking anything, no?

**Comment by MarcoFalke** [comment]
Side note that the fuzz tests are already upgraded to C++17: #18901 

I was wondering if we can also bump the msvc config to C++17 in advance, since the msvc build is primarily used by developers. And if someone used it in production, a recent msvc supporting C++17 should be available for all supported OSs. Any objections?

**Comment by laanwj** [comment]
Moving this to the 0.22 milestond, I don't think there's anything left to do here for 0.21.

**Comment by theuni** [comment]
Noting this here so it doesn't get lost....

I'm afraid it'll be quite some time before we can safely use c++17's `std::shared_mutex`.

When working on another project which uses read/write locks, we ran into this nasty glibc bug: https://sourceware.org/bugzilla/show_bug.cgi?id=23861.

Specifically the problem is that totally valid uses of `std::shared_mutex` may result in a deadlock deep in glibc.

Unfortunately, the shared mutex path (which we can't hit currently) was broken in glibc for quite a while. It was fixed here: https://sourceware.org/git/?p=glibc.git;a=commit;h=f21e8f8ca466320fed38bdb71526c574dae98026 in 2018.

Ubuntu (for ex) only got the backported fix this week: https://bugs.launchpad.net/ubuntu/+source/glibc/+bug/1864864

So until the world has moved on from this glibc bug, I'm afraid we'll need to pass on read/write locks :(

**Comment by MarcoFalke** [comment]
We should probably determine the version of libc that has this bug fixed (2.29?) and put that with a comment in the source code?

**Comment by ysangkok** [comment]
@theuni The bugzilla link you posted is for bug 23861. But the Ubuntu bug you posted links [23844](https://sourceware.org/bugzilla/show_bug.cgi?id=23844) (a different bug number). I am letting you know since I found it confusing. @MarcoFalke Bug 23844 was [deferred from 2.29 according to release notes](https://sourceware.org/glibc/wiki/Release/2.29#Deferred_to_next_release).

**Comment by ilyapopov** [comment]
This Ubuntu bug is about Ubuntu 18.04 Bionic. 

Newer versions of Ubuntu, such as 20.04 LTS and 20.10 already have had the fixed version from the beginning.

**Comment by luke-jr** [comment]
What about other distros? Typically RHEL is our limiting distro...

**Comment by MarcoFalke** [comment]
Closing this for now. If discussions about specific features are needed, they should go to #20419 or #19183 or whatever pull request proposes them.

**Comment by MarcoFalke** [comment]
std::fs discussion here: #20460

**Comment by ysangkok** [comment]
@MarcoFalke How old do the patched glibc version need to be before it can be assumed present? Glibc 2.30 which contains fixes for bugs 23861 and 23844 was released in August 2019, and backports are available. The bugs only affect users running old distros without updating.

**Comment by fanquake** [comment]
> The bugs only affect users running old distros without updating.

Ubuntu Bionic (18.04LTS) ships with [glibc 2.27](https://packages.ubuntu.com/source/bionic/glibc) and will remain supported until April 2023.

**Comment by ysangkok** [comment]
> Bionic (18.04LTS) ships glibc 2.27

@fanquake Yes, but it received a backported fix as shown in the Launchpad link. All users have `-updates` enabled by default, so the only way you could miss it is if you didn't run `apt upgrade` since November. Does Core need to cater to that tiny minority that runs an LTS release but does not keep it updated? You can only use Bitcoin if you are internet-connected anyway, so what is the rationale for being internet-connected, but keeping your system buggy and vulnerable (other fixes could be more severe)?
